x/metadata/client/cli/helpers.go

```diff
@@ -0,0 +1,309 @@
+package cli
+
+import (
+	"fmt"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/x/metadata/types"
+)
+
+// parseProcess parses a comma delimited string into a process.
+// Expected format: "<Name>,<ProcessId>,<Method>"
+// If <ProcessId> is not a metadata address, it is a hash.
+func parseProcess(commaDelimitedString string) (types.Process, error) {
+	var rv types.Process
+	parts := strings.Split(commaDelimitedString, ",")
+	if len(parts) != 3 {
+		return rv, fmt.Errorf("invalid process %q: expected 3 parts, have: %d", commaDelimitedString, len(parts))
+	}
+	rv.Name = parts[0]
+	_, err := types.MetadataAddressFromBech32(parts[1])
+	if err != nil {
+		rv.ProcessId = &types.Process_Hash{Hash: parts[1]}
+	} else {
+		rv.ProcessId = &types.Process_Address{Address: parts[1]}
+	}
+	rv.Method = parts[2]
+	return rv, nil
+}
+
+// parseParties parses a semicolon delimited string into a list of parties.
+// See also: parseParty.
+func parseParties(semicolonDelimitedString string) ([]types.Party, error) {
+	if len(semicolonDelimitedString) == 0 {
+		return nil, nil
+	}
+	inputs := strings.Split(semicolonDelimitedString, ";")
+	rv := make([]types.Party, len(inputs))
+	var err error
+	for i, input := range inputs {
+		rv[i], err = parseParty(input)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return rv, nil
+}
+
+// parseParty parses a comma delimited string into a Party.
+// Expected formats: "<address>" or "<address>,<party type>" or "<address>,<party type>,opt"
+// If no <party type> is provided, OWNER is used.
+// If "opt" is provided, the result has optional=true, otherwise optional=false.
+// See also: parsePartyType.
+func parseParty(commaDelimitedString string) (rv types.Party, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("invalid party %q: %w", commaDelimitedString, err)
+		}
+		if rv.Role == types.PartyType_PARTY_TYPE_UNSPECIFIED {
+			rv.Role = types.PartyType_PARTY_TYPE_OWNER
+		}
+	}()
+
+	if len(commaDelimitedString) == 0 {
+		return rv, fmt.Errorf("cannot be empty")
+	}
+	parts := strings.Split(commaDelimitedString, ",")
+	if len(parts) > 3 {
+		return rv, fmt.Errorf("expected 1, 2, or 3 parts, have: %d", len(parts))
+	}
+	_, err = sdk.AccAddressFromBech32(parts[0])
+	if err != nil {
+		return rv, fmt.Errorf("invalid address %q: %w", parts[0], err)
+	}
+	rv.Address = parts[0]
+	if len(parts) >= 2 {
+		// Don't change the role from default unless we know it parsed correctly.
+		rv.Role, err = parsePartyType(parts[1])
+		if err != nil {
```

Comment:
I had to change this back.

That second part needs to only happen when role parsing fails. So I could have it as `if err != nil && len(parts) != 2 { return rv, err }`, but then I'd still need an `if err != nil { ... }` around the second part.

Here's what I had changed it to, and why it didn't work:
```golang
if len(parts) >= 2 {
	// Don't change the role from default unless we know it parsed correctly.
	rv.Role, err = parsePartyType(parts[1])
	if err != nil && len(parts) != 2 {
		return rv, err
	}
	// allow for "<address>,opt"
	var optErr error
	rv.Optional, optErr = parseOptional(parts[1])
	if optErr != nil {
		// If the second (and only) thing wasn't an optional flag either, return the role error.
		return rv, err
	}
}
```

Consider the input string `"cosmos1v9jxgun9wde47v2lta047h6lta047h6lhgvprx,servicer,opt"`
During the parsing of `parts[1]` (`"servicer"`), it correctly parses it into that role and `err` is nil. It then continues on to try to parse "servicer" as an optional, but fails. Since that fails, it then returns the original error, which was nil. That stops it from moving on to try to parse the third part, though.

All that being said, I do like putting that first return closer to where the error is generated. I just still need the second part on level in, under an `if err != nil {...}`.